src/Network/Bugsnag/BeforeNotify.hs

```diff
@@ -180,13 +180,13 @@ redactRequestHeaders headers event =
 -- |
 --
 -- >>> let headers = [("Authorization", "secret"), ("X-Foo", "Bar")]
--- >>> let req = bugsnagRequest { brHeaders = Just headers }
+-- >>> let req = bugsnagRequest { brHeaders = Just $ BugsnagRequestHeaders headers }
 -- >>> brHeaders $ redactHeaders ["Authorization"] req
--- Just [("Authorization","<redacted>"),("X-Foo","Bar")]
+-- Just (BugsnagRequestHeaders {unBugsnagRequestHeaders = [("Authorization","<redacted>"),("X-Foo","Bar")]})
 --
 redactHeaders :: [HeaderName] -> BugsnagRequest -> BugsnagRequest
 redactHeaders headers request = request
-    { brHeaders = map redactHeader <$> brHeaders request
+    { brHeaders = BugsnagRequestHeaders . map redactHeader . unBugsnagRequestHeaders <$> brHeaders request
```

Comment:
Ideally we wouldn't need to export the constructor or accessor outside of the module containing this type, that way we can mess with its internals safely.

Can we:

- Define `BugsnagRequestHeaders` in its own module
- Don't export the constructor or field accessor
- Define a constructor function from `[Header]`
- Define a `redact :: [HeaderName] -> BugsnagRequestHeaders -> BugsnagRequestHeaders` that does this

Then update the above to:

```hs
redactRequestHeaders :: [HeaderName] -> BeforeNotify
redactRequestHeaders headers event =
    let request = beRequest event
        requestHeaders = BugsnagRequestHeaders.redact headers $ brHeaders request
    in event { beRequest = request { brHeaders = requestHeaders } }
```

WDYT?

Reply:
I don't mind doing it. I would propose these changes to your plan:

1. Also export a function `BugsnagRequestHeaders -> Headers` function. Being limited to just `redact` would be sad—ideally users could implement their own redact function for example. 
2. Export `BugsnagRequestHeaders` with the constructor in some sort of Internals module that isn't promising a backwards compatible API. This gives users more flexibility if they don't want to opt-out of API breaking guarantees

This would allow the internal representation of the headers to change still (e.g. it could be a `Map Text Text`, since that can easily become `Headers`).

Reply:
I'm good with (1) (Just export the `unBugsnagRequestReaders` accessor, right?). But I'm not particularly into (2). I don't think we need that yet. We'll have a constructor and an accessor exported at that point, so I can't really see what else a `.Internal` module would even be exporting.

Reply:
It would let you use things like `coerce` or derive additional instances, which might be nice for people

Reply:
Gotcha. Let's hold off on that for now. If this project were a big `.Internal` user already, sure, but I think we can wait for a concrete need to jump to that kind of organization.

Reply:
Ok, something like what I just pushed?